ide style completions

maxomatic458 opened the pull request:
#660 

demo clip: [![asciicast](https://asciinema.org/a/urUsZ5DbRieBEY27mxfDTgd9S.svg)](https://asciinema.org/a/urUsZ5DbRieBEY27mxfDTgd9S)

ysthakur commented:
This is so cool! Would it be possible to also show descriptions for each value when that particular value is highlighted? (like how IDEs show you doc comments when you go to a particular function)

maxomatic458 commented:
yes im planing to implement it like [inshellisense](https://github.com/microsoft/inshellisense)
so a description box to the right (or the left depending on how much space is available)

fdncred commented:
wow, that looks super cool! nice work!

maxomatic458 commented:
is there an easy way to add descriptions to a completer?
i cant find any way to do so. is that not implemented yet?

fdncred commented:
> is there an easy way to add descriptions to a completer? i cant find any way to do so. is that not implemented yet?

Seems like one of the types of completers has `value` and `description`.

![image](https://github.com/nushell/reedline/assets/343840/5ac6c85f-acc7-48c2-9154-8c6173bb4162)

This is how I've enabled this in nushell
### menu
```
  {
    name: commands_menu
    only_buffer_difference: false
    marker: "# "
    type: {
      layout: columnar
      columns: 4
      col_width: 20
      col_padding: 2
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      scope commands
      | where name =~ $buffer
      | each { |it| {value: $it.name description: $it.usage} }
    }
  }
```
### keybinding
```
  {
    name: commands_menu
    modifier: control
    keycode: char_t
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: commands_menu }
  }
```

maxomatic458 commented:
that seems to be nushell specific.
reedlines default completer doesnt seem to have that implemented

i guess for testing purposes i can just add the descriptions in the menu

fdncred commented:
I don't think nushell could have a menu on its own if reedline didn't support it. You can see here that the `Suggestion` struct supports a value and description.

https://github.com/nushell/reedline/blob/aa101f4d697dfebc886cd233e6a92d8957c27ca1/src/completion/base.rs#L56-L71

Nushell has this "convert_to_suggestion" function that may help if there's no example of how to do it in reedline itself.
https://github.com/nushell/nushell/blob/42bb42a2e1718469133fe3a71ac76c40389bbb30/crates/nu-cli/src/menus/menu_completions.rs#L76-L162

maxomatic458 commented:
yes i have seen this, but the DefaultCompleter in reedline always returns None as description
https://github.com/nushell/reedline/blob/aa101f4d697dfebc886cd233e6a92d8957c27ca1/src/completion/default.rs#L99-L105

reedline does support it, but the DefaultCompleter doesnt
so im just inserting some dummy values in the complete function for testing

maxomatic458 commented:
so i implemented descriptions like this
[![asciicast](https://asciinema.org/a/SSixmcIH5Fs7oAlE3E4YXPTIf.svg)](https://asciinema.org/a/SSixmcIH5Fs7oAlE3E4YXPTIf)

i made a enum to control where the description box shows up
https://github.com/maxomatic458/reedline/blob/94b7ebe829249c0ee3699b527ff90308bf7a258d/src/menu/ide_menu.rs#L9-L17

i also added an optional border
[![asciicast](https://asciinema.org/a/7X72M2h3XolA4fPvdkqWXXs5T.svg)](https://asciinema.org/a/7X72M2h3XolA4fPvdkqWXXs5T)

maxomatic458 commented:
should i add checks to make sure that the description or completion has a valid size? (maybe cut off some stuff if it doesnt fit)
Im asking because the other menus dont have any checks for that

fdncred commented:
I thought there was a limit or it wrapped or something?

maxomatic458 commented:
i added that for the description box currently
but not for the completions

i guess ill cut of strings that dont fit.
the list & column menus dont seem to check for that

fdncred commented:
maybe truncate with `...` indicating that they were truncated?

fdncred commented:
@maxomatic458 Do you have time to clean up these CI errors?

maxomatic458 commented:
@fdncred i will once im done, i marked it as draft again, i decided i want to change some stuff (more user input validation)

fdncred commented:
Ah, ok. We just made a nushell release and we like to land things like this early after a release to give us time to test it. Not rushing you, just explaining how we usually work.

fdncred commented:
yay, it's green. are you happy with what you have here?

maxomatic458 commented:
@fdncred yep, should be done now

fdncred commented:
This PR is pretty benign. It's adding something new. I'm anxious to see a PR in nushell to enable this somehow. testing this now but I don't see an example with description. Is there something I have to do to enable that or is there just not one yet? Also, maybe I'm just missing them, but I don't see any tests. We've found that if you want to not break something in the future, it's good to have tests that demonstrate positive and negative responses. If you can add some tests and and example with descriptions (assuming that still works), i'll land this PR.

maxomatic458 commented:
@fdncred i didnt find a way to add completions in reedline alone. for testing i manually inserted some here https://github.com/nushell/reedline/blob/0c5f9817465df3c80e062dd1b5b05bdd91924b83/src/completion/default.rs#L102-L108
and i also tested this menu in nushell.

the completion logic is basically just copied over from the column menu (wouldn't make sense to add tests here)

i could add a test for this function https://github.com/maxomatic458/reedline/blob/d4c00e0859bbcbe0429753b02c9b4adbf10a9310/src/menu/ide_menu.rs#L1040-L1108

but not really sure where else i could add them.
the only real other thing that could fail is that the menu doesnt render correctly and im not sure how this could be tested properly

fdncred commented:
if you don't mind, we'd like to have a much tested as possible, even if it's just a few things. the more testing the more stable the feature will be.

maxomatic458 commented:
should i copy over the completion tests from the column menu as well (and just replace the menu with the ide menu)? i didnt touch the completion logic, but i guess it probably doesnt hurt to have them.

fdncred commented:
> should i copy over the completion tests from the column menu as well

if that works and exercises your new menu, i'm fine with it.

ysthakur commented on src/menu/ide_menu.rs:
```diff
@@ -0,0 +1,1108 @@
+    }
+
+    /// Update menu values
+    fn update_values(&mut self, editor: &mut Editor, completer: &mut dyn Completer) {
```
By the way, the columnar menu was [recently updated](https://github.com/nushell/reedline/commit/0c5f9817465df3c80e062dd1b5b05bdd91924b83#diff-3fc955a6de109c1bf50c491b78e1794fcff831e06bba9d4f2455c8d68e5fba1d) so that `only_buffer_difference: true` would work, so you should probably copy the `update_values` method over again. Can't wait to use this!

fdncred commented:
looking good. ping me when you're ready to land this. i'm anxious to play with it in nushell.

maxomatic458 commented:
@fdncred i think im ready. [here](https://github.com/maxomatic458/nushell) is my nushell fork where you can try out the new menu.

the reedline example still doesn't include descriptions, as that would require either changing the default completer, or adding a new one. which should probably be done in a seperate PR.

maxomatic458 commented:
here is an example menu entry for the config.nu file
```
    {
      name: completion_menu
      only_buffer_difference: false
      marker: "| "
      type: {
          layout: ide
          border: true
      }
      style: {
          text: green
          selected_text: green_reverse
          description_text: yellow
      }
    }
```

fdncred commented:
I tried your branch out and the completion list jumps around. I'm wondering if it has something to do with my prompt indicator being on a new line or something?
![jumping2](https://github.com/nushell/reedline/assets/343840/f9c61c3e-103c-4fd1-8497-aaf2539f90cd)

maxomatic458 commented:
@fdncred hmm...
Shouldn't be caused by that (mine is also on a new line), ill try to see if i can replicate this

fdncred commented:
if you watch the gif, i did notice that the `| str` starts on the prompt line instead of on the indicator line. all completions may do that, not sure, i just noticed it here.

if it's any help, this is how my prompt is setup
```nushell
# prompt
use modules\prompt\oh-my.nu git_prompt
$env.PROMPT_COMMAND = {|| (git_prompt).left_prompt }
$env.PROMPT_COMMAND_RIGHT = {|| (git_prompt).right_prompt }
$env.PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi green_bold)\n❯ (ansi reset)"} else {$"(ansi red_bold)\n❯ (ansi reset)"}}
$env.TRANSIENT_PROMPT_COMMAND = ""
$env.TRANSIENT_PROMPT_COMMAND_RIGHT = {|| (git_prompt).right_prompt }
$env.TRANSIENT_PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi light_yellow_bold)❯ (ansi reset)"} else {$"(ansi light_red_bold)❯ (ansi reset)"}}
```

and the `use` path above points to my local nu_scripts repo.

maxomatic458 commented:
@fdncred 

im experiencing the same problem when using the latest version of nushell (the official repo, with a columnar menu and your prompt settings).
Maybe its because of https://github.com/nushell/nushell/pull/11488

fdncred commented:
maybe? @ysthakur any thoughts on this?

ysthakur commented:
@fdncred I think the reason the `| str` isn't on a new line is because the `marker` for your completion menu doesn't have a newline preceding it. Does your columnar menu's config have `marker: "\n| "` rather than `"| "` maybe?

As for the jumping around, I don't think https://github.com/nushell/nushell/pull/11488 would cause a change to the prompt, only to the returned suggestions (and their spans), but it's possible something I did in #700 messes with the way the prompt is printed (although the changes there should only be applicable when the buffer can't fit on your screen). I'll investigate further later today

maxomatic458 commented:
``marker: "\n| "``
will show it on the correct line, but without any marker at all.
anything after the newline seems to get ignored.

ysthakur commented:
@maxomatic458 Ah, I think that's because when figuring out the starting row to print the menu at, the painter doesn't consider that the menu's indicator may have newlines (but the prompt indicator is taken into account). In that case, I don't know how fdncred's config manages to keep the prompt on the second line when completing with a regular columnar menu.

maxomatic458 commented:
so this cursor jumping when triggering a completion is not intended?
so should i wait until this gets sorted out?

fdncred commented:
I'm less concerned about `| ls` showing on the prompt vs indicator line but I'm more concerned about the entire menu jumping from the left to the right. That's the problem we need to solve.

fdncred commented:
@ysthakur I'm just using the suggested ide menu here https://github.com/nushell/reedline/pull/696#issuecomment-1889780555 and using @maxomatic458's nushell repo here https://github.com/nushell/reedline/pull/696#issuecomment-1889777321

ysthakur commented:
@fdncred I was curious what configuration you normally use for your menu, because I tried copying your prompt configuration and using it with the regular columnar menu, and I still got the issue with `| ls` showing up on the first line instead of the second, with the second line being blank. I realize now it's probably not related to the jumping issue though

fdncred commented:
Here's my config.nu. Lots of debug and testing stuff.
```nushell
# version output to markdown
def nuvermd [] { version | transpose key value | to md }
# version output to nushell table
def nuver [] { version | transpose key value }
# use pbcopy on macos to copy to clipboard
alias clip = pbcopy
# ls with colors, icons and grid
def lsg [] { ls | sort-by type name -i | grid -c }
# figure out how many lines are in code with tokei
def tok [] { 
    tokei | lines | skip 1 | str join "\n" | detect columns | 
    where Language !~ "=" and Language !~ "-" and Language !~ '\(' |
    into int Files Lines Code Comments Blanks
}

#alias for lazygit
alias lg = lazygit

# alias lvim = ^/Users/fdncred/.local/bin/lvim
# exec -a "$NVIM_APPNAME" nvim -u "$LUNARVIM_BASE_DIR/init.lua" "$@"

def path_underline [color: string] {
    {|p| if ($p | path exists) {
            $color + "_underline"
        } else {
            $color
        }
    }
}

def is_alias [cmd: string] {
    $cmd in (scope aliases | get name)
}

# # abbreviation for git status
# alias gst = git status
# # List all local branches
# alias gb = git branch
# Fix vi mistakes
alias vi = nvim
# Fix vim mistakes
alias vim = nvim

# alias gfu = git fetch upstream
# alias gmum = git merge upstream/main
# alias gpom = git push origin main
# alias gcom = git checkout main

# long listing
alias ll = ls -l

# region: Themes

# attributes
# code	meaning
# l	    blink
# b	    bold
# d	    dimmed
# h	    hidden
# i	    italic
# r	    reverse
# s	    strikethrough
# u	    underline
# n	    nothing
# defaults to nothing

# let base00 = "#181818" # Default Background
# let base01 = "#282828" # Lighter Background (Used for status bars, line number and folding marks)
# let base02 = "#383838" # Selection Background
# let base03 = "#585858" # Comments, Invisibles, Line Highlighting
# let base04 = "#b8b8b8" # Dark Foreground (Used for status bars)
# let base05 = "#d8d8d8" # Default Foreground, Caret, Delimiters, Operators
# let base06 = "#e8e8e8" # Light Foreground (Not often used)
# let base07 = "#f8f8f8" # Light Background (Not often used)
# let base08 = "#ab4642" # Variables, XML Tags, Markup Link Text, Markup Lists, Diff Deleted
# let base09 = "#dc9656" # Integers, Boolean, Constants, XML Attributes, Markup Link Url
# let base0a = "#f7ca88" # Classes, Markup Bold, Search Text Background
# let base0b = "#a1b56c" # Strings, Inherited Class, Markup Code, Diff Inserted
# let base0c = "#86c1b9" # Support, Regular Expressions, Escape Characters, Markup Quotes
# let base0d = "#7cafc2" # Functions, Methods, Attribute IDs, Headings
# let base0e = "#ba8baf" # Keywords, Storage, Selector, Markup Italic, Diff Changed
# let base0f = "#a16946" # Deprecated, Opening/Closing Embedded Language Tags, e.g. <?php ?>

# let base16_theme = {
#   separator: $base03
#   leading_trailing_space_bg: $base04
#   header: $base0b
#   date: $base0e
#   filesize: $base0d
#   row_index: $base0c
#   bool: $base08
#   int: $base0b
#   duration: $base08
#   range: $base08
#   float: $base08
#   string: $base04
#   nothing: $base08
#   binary: $base08
#   cellpath: $base08

#   # shape_garbage: { fg: $base07 bg: $base08 attr: b} # base16 white on red
#   # but i like the regular white on red for parse errors
#   shape_garbage: { fg: "#FFFFFF" bg: "#FF0000" attr: b}
#   shape_bool: $base0d
#   shape_int: { fg: $base0e attr: b}
#   shape_float: { fg: $base0e attr: b}
#   shape_range: { fg: $base0a attr: b}
#   shape_internalcall: { fg: $base0c attr: b}
#   shape_external: $base0c
#   shape_externalarg: { fg: $base0b attr: b}
#   shape_literal: $base0d
#   shape_operator: $base0a
#   shape_signature: { fg: $base0b attr: b}
#   shape_string: $base0b
#   shape_filepath: $base0d
#   shape_globpattern: { fg: $base0d attr: b}
#   shape_variable: $base0e
#   shape_flag: { fg: $base0d attr: b}
#   shape_custom: {attr: b}
#   shape_matching_brackets: { attr: u }
# }

let my_theme = {
  binary: red
  bool: {|| if $in { 'light_cyan' } else { 'light_red' } }
  cellpath: cyan
  date: {|| (date now) - $in |
    if $in < 1hr {
      'red3b' #"\e[38;5;160m" #'#e61919' # 160
    } else if $in < 6hr {
      'orange3' #"\e[38;5;172m" #'#e68019' # 172
    } else if $in < 1day {
      'yellow3b' #"\e[38;5;184m" #'#e5e619' # 184
    } else if $in < 3day {
      'chartreuse2b' #"\e[38;5;112m" #'#80e619' # 112
 
      } else if $in < 1wk {
      'green3b' #"\e[38;5;40m" #'#19e619' # 40
    } else if $in < 6wk {
      'darkturquoise' #"\e[38;5;44m" #'#19e5e6' # 44
    } else if $in < 52wk {
      'deepskyblue3b' #"\e[38;5;32m" #'#197fe6' # 32
    } else { 'dark_gray' }
  }

  duration: blue_bold
  filesize: {|e| if $e == 0b { 'white' } else if $e < 1mb { 'blue' } else { 'cyan' } }
  float: red
  header: gb
  hints: dark_gray
  int: green
  leading_trailing_space_bg: {bg: dark_gray_dimmed}
  nothing: red
  range: purple
  row_index: cb
  separator: purple
#   string: { if $in =~ '^#\w{6}$' { $in } else { 'white' } }
  string: {|| if $in =~ '^#[a-fA-F\d]+' { $in } else { 'white' } }
  search_result: {bg: red fg: white}
#   search_result: blue_reverse

  shape_and: purple_bold
  shape_binary: purple_bold
  shape_block: "#33ff00"
  shape_bool: light_cyan
  shape_closure: "#ffb000"
  shape_custom: green
  shape_datetime: cyan_bold
  shape_directory: cyan
  shape_external: darkorange
  shape_externalarg: green_bold
  shape_external_resolved: light_yellow_bold
  shape_filepath: cyan
  shape_flag: blue_bold
  shape_float: purple_bold
  # shapes are used to change the cli syntax highlighting
  shape_garbage: { fg: white bg: red attr: b}
  shape_globpattern: cyan_bold
  shape_int: purple_bold
  shape_internalcall: cyan_bold
  shape_keyword: cyan_bold
  shape_list: cyan_bold
  shape_literal: blue
  shape_match_pattern: green
  shape_matching_brackets: { attr: u }
#    shape_matching_brackets: { fg: red bg: default attr: b }
  shape_nothing: light_cyan
  shape_operator: yellow
  shape_or: purple_bold
  shape_pipe: purple_bold
  shape_range: yellow_bold
  shape_record: cyan_bold
  shape_redirection: purple_bold
  shape_signature: green_bold
  shape_string: green
  shape_string_interpolation: cyan_bold
  shape_table: blue_bold
  shape_variable: purple
  shape_vardecl: purple
}

# For more information on themes, see
# https://www.nushell.sh/book/coloring_and_theming.html
let dark_theme = {
    # color for nushell primitives
    separator: white
    leading_trailing_space_bg: { attr: n } # no fg, no bg, attr none effectively turns this off
    header: green_bold
    empty: blue
    # Closures can be used to choose colors for specific values.
    # The value (in this case, a bool) is piped into the closure.
    bool: {|| if $in { 'light_cyan' } else { 'light_red' } }
    int: white
    filesize: {|e|
      if $e == 0b {
        'white'
      } else if $e < 1mb {
        'cyan'
      } else { 'blue' }
    }
    duration: white
    date: {|| (date now) - $in |
      if $in < 1hr {
        'red3b'
      } else if $in < 6hr {
        'orange3'
      } else if $in < 1day {
        'yellow3b'
      } else if $in < 3day {
        'chartreuse2b'
      } else if $in < 1wk {
        'green3b'
      } else if $in < 6wk {
        'darkturquoise'
      } else if $in < 52wk {
        'deepskyblue3b'
      } else { 'dark_gray' }
    }
    range: white
    float: white
    string: white
    nothing: white
    binary: white
    cellpath: white
    row_index: green_bold
    record: white
    list: white
    block: white
    hints: dark_gray

    shape_and: purple_bold
    shape_binary: purple_bold
    shape_block: blue_bold
    shape_bool: light_cyan
    shape_custom: green
    shape_datetime: cyan_bold
    shape_directory: cyan
    shape_external: cyan
    shape_externalarg: green_bold
    shape_filepath: cyan
    shape_flag: blue_bold
    shape_float: purple_bold
    # shapes are used to change the cli syntax highlighting
    shape_garbage: { fg: "#FFFFFF" bg: "#FF0000" attr: b}
    shape_globpattern: cyan_bold
    shape_int: purple_bold
    shape_internalcall: cyan_bold
    shape_list: cyan_bold
    shape_literal: blue
    shape_match_pattern: green
    shape_matching_brackets: { attr: u }
    shape_nothing: light_cyan
    shape_operator: yellow
    shape_or: purple_bold
    shape_pipe: purple_bold
    shape_range: yellow_bold
    shape_record: cyan_bold
    shape_redirection: purple_bold
    shape_signature: green_bold
    shape_string: green
    shape_string_interpolation: cyan_bold
    shape_table: blue_bold
    shape_variable: purple
}
# endregion

# region: External Completers
# let external_completer = {|spans| 
#   {
#     $spans.0: {carapace $spans.0 nushell $spans | from json } # default
#   } | get $spans.0 | each {|it| do $it}
# }

# let external_completer = {|spans|
# {
#   $spans.0: {carapace $spans.0 nushell $spans | from json } # default
#   "empty": { "[]" | from json } # no result (valid json)
#   "unknown": { "" | from json } # default file completion (invalid json)
#   "vals": { '["a", "b", "c"]' | from json } # external completion (valid json)
#   } | get $spans.0 | each {|it| do $it}
# }
# let fish_completer = {|spans|
#   fish --command $'complete "--do-complete=($spans | str join " ")"' 
#   | str trim | split row "\n" | each { |line| $line | split column "\t" value description } | flatten
# }

# latest external completers from the cookbook 7/18/2023
# let carapace_completer = {|spans|
#     carapace $spans.0 nushell $spans | from json
# }

let fish_completer = {|spans|
    fish --command $'complete "--do-complete=($spans | str join " ")"'
    | $"value(char tab)description(char newline)" + $in
    | from tsv --flexible --no-infer
}

let zoxide_completer = {|spans|
    $spans | skip 1 | zoxide query -l $in | lines | where {|x| $x != $env.PWD}
}

let carapace_completer = {|spans: list<string>|
    carapace $spans.0 nushell $spans
    | from json
    | if ($in | default [] | where value =~ '^-.*ERR$' | is-empty) { $in } else { null }
}

let multiple_completers = {|spans|
    let expanded_alias = (scope aliases | where name == $spans.0 | get -i 0 | get -i expansion)

    let spans = (if $expanded_alias != null  {
        $spans | skip 1 | prepend ($expanded_alias | split words)
    } else { $spans })

    # match spans.0 {
    #     "z" | "zi" => $zoxide_completer
    #     "nu" => $fish_completer
    #     "git" => $carapace_completer
    #     _ => $carapace_completer
    # }
    
    {
      # zoxide alias
      z: $zoxide_completer

      # zoxide alias
      zi: $zoxide_completer

      # carapace completions are incorrect for nu
      nu: $fish_completer

      # fish completes commits and branch names in a nicer way
      git: $fish_completer
    } | get -i $spans.0 | default $carapace_completer | do $in $spans # use carapace as default
}

# endregion

# region: Config
$env.config = {
  show_banner: true # true or false to enable or disable the banner

  ls: {
    use_ls_colors: true # use the LS_COLORS environment variable to colorize output
    clickable_links: true # true or false to enable or disable clickable links in the ls listing. your terminal has to support links.
  }

  rm: {
    always_trash: true # always act as if -t was given. Can be overridden with -p
  }

  table: {
    mode: rounded # basic, compact, compact_double, light, thin, with_love, rounded, reinforced, heavy, none, other
    index_mode: always #  "always" show indexes, "never" show indexes, "auto" = show indexes when a table has "index" column
    show_empty: false # show 'empty list' and 'empty record' placeholders for command output
    trim: {     # A strategy of managing table view in case of limited space.
      methodology: wrapping # truncating or wrapping
      wrapping_try_keep_words: true
      # A suffix which will be used with 'truncating' methodology
      truncating_suffix: "…" # "⋯" "︙"
    }
    header_on_separator: true # show header text on separator/border line
    padding: {left: 1, right: 1} # int or record {left: 0 right: 0}
  }
  history: {
    max_size: 1_000_000 # Session has to be reloaded for this to take effect
    sync_on_enter: true # Enable to share the history between multiple sessions, else you have to close the session to persist history to file
    file_format: sqlite # "sqlite" or "plaintext"
    isolation: false # true enables history isolation, false disables it. history_isolation is a unique history per session.
  }
  completions: {
    case_sensitive: false # set to true to enable case-sensitive completions
    quick: true # set this to false to prevent auto-selecting completions when only one remains
    partial: true # set this to false to prevent partial filling of the prompt
    algorithm: "prefix" # prefix, fuzzy
    external: {
      enable: true # set to false to prevent nushell looking into $env.PATH to find more suggestions, `false` recommended for WSL users as this look up my be very slow
      max_results: 100 # setting it lower can improve completion performance at the cost of omitting some options
    #   completer: $multiple_completers
      completer: null
    }
  }
  filesize: {
    metric: true # true => KB, MB, GB (ISO standard), false => KiB, MiB, GiB (Windows standard)
    format: "auto" # b, kb, kib, mb, mib, gb, gib, tb, tib, pb, pib, eb, eib, zb, zib, auto
  }
  cursor_shape: {
    emacs: line
    vi_insert: block
    vi_normal: underscore
  }
#   datetime_format: {
#     # normal: "%Y-%m-%d %H:%M:%S"
#     normal: "%F %r"
#     # table: "%Y-%m-%d %H:%M:%S"
#     # table: "%b %d %Y %I:%M:%S %p"
#     # table: "%s"
#     # table: "%F %r"
#     table: "%x %r"
#   }

  datetime_format: {
    # normal: '%a, %d %b %Y %H:%M:%S %z'  # shows up in displays of variables or other datetime's outside of tables
    # table: '%m/%d/%y %I:%M:%S%p'        # generally shows up in tabular outputs such as ls
  }
  error_style: fancy
  color_config: $my_theme
  use_grid_icons: true
  footer_mode: 5 #always, never, number_of_rows, auto
  float_precision: 2
  buffer_editor: "hx" # command that will be used to edit the current line buffer with ctr+e
  use_ansi_coloring: true
  bracketed_paste: true # enable bracketed paste, currently useless on windows
  edit_mode: emacs # emacs, vi
  shell_integration: true # enables terminal markers and a workaround to arrow keys stop working issue
  render_right_prompt_on_last_line: false # true or false to enable or disable right prompt to be rendered on last line of the prompt.
  use_kitty_protocol: true # enables keyboard enhancement protocol implemented by kitty console, only if your terminal support this
  highlight_resolved_externals: true

  explore: {
    config: {
        cursor_color: 'red'
    }
    table: {
        selected_cell: { bg: 'blue'}
        show_cursor: false
    }
    try: {
        reactive: true
    }
  }

  hooks: {
    pre_prompt: [{||
      #if (shells | length) > 1 { tabline-hook }
      null
      #print -n "pre_p " (date now | format date "%Y-%m-%d %H:%M:%S.%f") (char nl)
      #if "LAST_CMD" in $env {
      #  print $"(ansi title)(pwd | str replace "/Users/fdncred" "~") > ($env.LAST_CMD)(ansi st)"
      #} else {
      #  print $"(ansi title)(pwd | str replace "/Users/fdncred" "~")(ansi st)"
      #}
    }]
    pre_execution: [{||
      null
      # $env.LAST_CMD = (commandline)
      #print -n "pre_e " (date now | format date "%Y-%m-%d %H:%M:%S.%f") (char nl)
    }]
    env_change: {
      PWD: [
      { |before, after|
        #print $"before: [($before)], after: [($after)]"
        # print ($env.LS_COLORS)
        print (lsg)
        # null
      },
      { 
        condition: {|_, after| not ($after | path join 'toolkit.nu' | path exists)}
        code: "hide toolkit"
        # code: "overlay hide --keep-env [ PWD ] toolkit"
      },
      {
        condition: {|_, after| $after | path join 'toolkit.nu' | path exists}
        code: "
        print $'(ansi default_underline)(ansi default_bold)toolkit(ansi reset) module (ansi green_italic)detected(ansi reset)...'
        print $'(ansi yellow_italic)activating(ansi reset) (ansi default_underline)(ansi default_bold)toolkit(ansi reset) module with `(ansi default_dimmed)(ansi default_italic)use toolkit.nu(ansi reset)`'
        use toolkit.nu
        # overlay use --prefix toolkit.nu
        "
      },
      {|before, _|
        if $before == null {
            let file = ($nu.home-path | path join ".local" "share" "nushell" "startup-times.nuon")
            if not ($file | path exists) {
                mkdir ($file | path dirname)
                touch $file
            }
            let ver = (version)
            open $file | append {
                date: (date now)
                time: $nu.startup-time
                build: ($ver.build_rust_channel)
                allocator: ($ver.allocator)
                version: ($ver.version)
                commit: ($ver.commit_hash)
                build_time: ($ver.build_time)
            } | save --force $file
        }
      }      
      ]
    }
    # display_output: {
    #   if (term size).columns > 100 { table -e } else { table }
    # #   table
    # }
    # display_output: {
    #   $env.__ = $in;
    #   print $env.__;
    # }
    display_output: { table } # run before the output of a command is drawn, example: `{ if (term size).columns >= 100 { table -e } else { table } }`
    command_not_found: {||
      null  # replace with source code to return an error message when a command is not found
    }
  }
  # old timey crt colors
  #   #ffb000 <- 600nm P3 <--history_menu
  #   #ffcc00 <- 593nm
  #   #33ff00 <- 524nm <-- completion_menu
  #   #33ff33 <- P1
  #   #00ff33 <- 506nm
  #   #66ff66 <- P24
  #   #00ff66 <- 502nm
  #   #282828 <- background

  # log_level: trace
  # region: Menus
  menus: [
  # Configuration for default nushell menus
  # Note the lack of souce parameter
  {
    name: completion_menu
    only_buffer_difference: false
    marker: $"(char -u '1f4ce') "
    type: {
      layout: columnar
      columns: 4
      col_width: 20   # Optional value. If missing all the screen width is used to calculate column width
      col_padding: 2
    }
    style: {
      text: { fg: "#33ff00" }
      selected_text: { fg: "#0c0c0c" bg: "#33ff00" attr: b}
      description_text: yellow
    }
  }
  {
    name: history_menu
    only_buffer_difference: true
    marker: $"(char -u '1f50d') "
    type: {
      layout: list
      page_size: 10
    }
    style: {
      text: "#ffb000"
      selected_text: { fg: "#ffb000" attr: r }
      description_text: yellow
    }
  }
  {
    name: help_menu
    only_buffer_difference: true
    marker: "? "
    type: {
      layout: description
      columns: 4
      col_width: 20   # Optional value. If missing all the screen width is used to calculate column width
      col_padding: 2
      selection_rows: 4
      description_rows: 10
    }
    style: {
      text: "#7F00FF"
      selected_text: { fg: "#ffff00" bg: "#7F00FF" attr: b }
      description_text: "#ffff00"
    }
  }
  # Example of extra menus created using a nushell source
  # Use the source field to create a list of records that populates
  # the menu
  {
    name: fzf_history_menu_fzf_ui
    only_buffer_difference: false
    marker: "# "
    type: {
      layout: columnar
      columns: 4
      col_width: 20
      col_padding: 2
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      open $nu.history-path | get history.command_line
      | to text | fzf +s --tac | str trim
      | where $it =~ $buffer
      | each { |v| {value: ($v | str trim) } }
    }
  }
  {
    name: fzf_menu_nu_ui
    only_buffer_difference: false
    marker: "# "
    type: {
      layout: list
      page_size: 10
    }
    style: {
      text: "#66ff66"
      selected_text: { fg: "#66ff66" attr: r }
      description_text: yellow
    }        
    source: { |buffer, position|
      open $nu.history-path | get history.command_line
      | to text
      | fzf -f $buffer
      | lines
      | each { |v| {value: ($v | str trim) } }
    }
  }
  {
    name: commands_menu
    only_buffer_difference: false
    marker: "# "
    type: {
      layout: columnar
      columns: 4
      col_width: 20
      col_padding: 2
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      scope commands
      | where name =~ $buffer
      | each { |it| {value: $it.name description: $it.usage} }
    }
  }
  {
    name: vars_menu
    only_buffer_difference: false
    marker: "👀 "
    type: {
      layout: columnar
      columns: 1
      col_width: 20
      col_padding: 2
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      scope variables
      | where name == $buffer
      | each { |it| {value: $it.value }}
    }
  }
  {
    name: commands_with_description
    only_buffer_difference: true
    marker: "# "
    type: {
      layout: description
      columns: 4
      col_width: 20
      col_padding: 2
      selection_rows: 4
      description_rows: 10
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      scope commands
      | where name =~ $buffer
      | each { |it| {value: $it.name description: $it.usage} }
    }
  }
  {
    name: alias_menu
    only_buffer_difference: false
    marker: "👀 "
    type: {
      layout: columnar
      columns: 1
      col_width: 20
      col_padding: 2
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      scope aliases
      | where name == $buffer
      | each { |it| {value: $it.expansion }}
    }
  }
  {
    name: grid_history_menu
    only_buffer_difference: true
    marker: "<?> "
    type: {
      layout: description
      columns: 4
      col_width: 20
      col_padding: 2
      selection_rows: 4
      description_rows: 10  
    }
    style: {
      text: green
      selected_text: green_reverse
      description_text: yellow
    }
    source: { |buffer, position|
      history 
      | where command =~ $buffer
      | where exit_status < 1
      | get command
      | each  { |it| {value: $it description: (try {help $"($it | str trim)"| lines | get 0} catch { "" }) } }
    }
  }
  ]
  # endregion
  # region: Keybindings
   # mac keyboard
   # home = fn + left
   # end = fn + right
   # pageup = fn + up
   # pagedown = fn + down
   # backspace = delete
   # delete = fn + delete

  keybindings: [
  {
    name: insert_newline
    modifier: alt
    keycode: char_i
    mode: [emacs vi_normal vi_insert]
    event: { edit: insertnewline }
  }
  {
    name: insert_newline
    modifier: alt
    keycode: enter
    mode: [emacs vi_normal vi_insert]
    event: { edit: insertnewline }
  }
#   {
#     name: insert_newline
#     modifier: shift
#     keycode: (char cr)
#     mode: [emacs vi_normal vi_insert]
#     event: { edit: insertnewline }
#   }
  {
    name: trigger-completion-menu
    modifier: none
    keycode: tab
    mode: [emacs vi_normal vi_insert]
    event: { 
      until: [
        { send: menu name: completion_menu }
        { send: menunext }
        { edit: complete }
      ]
    }
  }
  {
    name: completion_previous
    modifier: shift
    keycode: backtab
    mode: [emacs vi_normal vi_insert]
    event: { send: menuprevious }
  }
  {
    name: clear
    modifier: none
    keycode: esc
    mode: [emacs vi_normal vi_insert]
    event: { edit: clear }
  }
  { 
    name: complete_hint_chunk
    modifier: alt
    keycode: right
    mode: [emacs vi_normal vi_insert]
    event: { until: [ 
      { send: historyhintwordcomplete }
      { edit: movewordright }
    ]
    }
  }
  { 
    name: un_complete_hint_chunk
    modifier: alt
    keycode: left
    mode: [emacs vi_normal vi_insert]
    event: [ 
      { edit: backspaceword }
    ]
  }
  {
    name: trigger-history-menu
    modifier: control
    keycode: char_x
    mode: [emacs vi_normal vi_insert]
    event: { 
      until: [
        { send: menu name: history_menu }
        { send: menupagenext }
      ]
    }
  }
  {
    name: trigger-history-previous
    modifier: control
    keycode: char_z
    mode: [emacs vi_normal vi_insert]
    event: { until: [
      { send: menupageprevious }
      { edit: undo }
      ]
    }
  }
  {
    name: trigger-help-menu
    modifier: control
    keycode: char_q
    mode: [emacs vi_normal vi_insert]
    event: { 
      until: [
        { send: menu name: help_menu }
        { send: menunext }
      ]
    }
  }  
  {
    name: prev_shell
    modifier: control
    keycode: char_p
    mode: [emacs vi_normal vi_insert]
    event: [
      { edit: clear }
      { edit: insertchar value: p }
      { send: enter }
    ]
  }
  {
    name: next_shell
    modifier: control
    keycode: char_n
    mode: [emacs vi_normal vi_insert]
    event: [
      { edit: clear }
      { edit: insertchar value: n }
      { send: enter }
    ]
  }
  # {
  #   name: change_dir_with_fzf
  #   modifier: control
  #   keycode: char_f
  #   mode: [emacs vi_normal vi_insert]
  #   event: {
  #     send: executehostcommand,
  #     cmd: "cd (ls | where type == dir | each { |it| $it.name} | str join (char nl) | fzf | decode utf-8 | str trim)"
  #   }
  # }
  {
    name: complete_in_cd
    modifier: none
    keycode: f2
    mode: [emacs vi_normal vi_insert]
    event: [
      { edit: clear }
      { edit: insertString value: "./" }
      { send: Menu name: completion_menu }
    ]
  }
  {
    name: reload_config
    modifier: none
    keycode: f5
    mode: [emacs vi_normal vi_insert]
    event: {
      send: executehostcommand,
      cmd: $"source '($nu.env-path)';source '($nu.config-path)'"
    }
  }
  {
    name: clear_backbuffer
    modifier: control
    keycode: char_l
    mode: [emacs vi_normal vi_insert]
    event: { send: ClearScrollback }
  }
  # {
  #     name: edit_buffer
  #     modifier: control
  #     keycode: char_i
  #     mode: [emacs vi_normal vi_insert]
  #     event: { send: OpenEditor }
  # }
  {
    name: insert_last_token
    modifier: alt
    keycode: char_.
    mode: [emacs vi_normal vi_insert]
    event: [
      { edit: InsertString, value: " !$" }
      { send: Enter }
    ]
  }
  # Keybindings used to trigger the user defined menus
  { #notworking
    name: fzf_history_menu_fzf_ui
    modifier: alt
    keycode: char_e
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: fzf_history_menu_fzf_ui }
  }
  {
    name: fzf_menu_nu_ui
    modifier: alt
    keycode: char_w
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: fzf_menu_nu_ui }
  }
  {
    name: commands_menu
    modifier: control
    keycode: char_t
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: commands_menu }
  }
  {
    name: vars_menu
    modifier: control
    keycode: char_y
    mode: [emacs, vi_normal, vi_insert]
    event:[ 
      { send: menu name: vars_menu }
      { edit: insertchar, value: ' '}
    ]
  }
  {
    name: commands_with_description
    modifier: control
    keycode: char_u
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: commands_with_description }
  }
  {
    name: alias_menu
    modifier: control
    keycode: char_a
    mode: [emacs, vi_normal, vi_insert]
    event: { send: menu name: alias_menu }
  }
#   {
#     name: fuzzy_history_skim
#     modifier: control
#     keycode: char_r
#     mode: [emacs vi_normal vi_insert]
#     event: {
#       send: executehostcommand
#       cmd: "commandline (history | each { |it| $it.command } | uniq | reverse | str join (char nl) | sk --layout=reverse --height=40% -q (commandline) | decode utf-8 | str trim)"
#     }
#   }
  {
    name: insert_last_arg_from_prev_cmd
    modifier: alt
    keycode: char_.
    mode: [emacs, vi_normal, vi_insert]
    event: {
        send: executeHostCommand
        cmd: "commandline --insert (history | last | get command | parse --regex '(?P<arg>[^ ]+)$' | get arg | first)"
    }
  }
  {
    name: fuzzy_history_fzf
    modifier: control
    keycode: char_r
    mode: [emacs , vi_normal, vi_insert]
    event: {
      send: executehostcommand
      cmd: "commandline (
        history 
          | each { |it| $it.command } 
          | uniq 
          | reverse 
          | str join (char -i 0) 
          | fzf --read0 --tiebreak=chunk --layout=reverse  --multi --preview='echo {..}' --preview-window='bottom:3:wrap' --bind alt-up:preview-up,alt-down:preview-down --height=70% -q (commandline) 
          | decode utf-8 
          | str trim
      )"
    }
  }
#   {
#     name: fuzzy_history
#     modifier: control
#     keycode: char_r
#     mode: [emacs, vi_normal, vi_insert]
#     event: [
#       {
#         send: ExecuteHostCommand
#         cmd: "commandline (
#           history
#             | each { |it| $it.command }
#             | uniq
#             | reverse
#             | str join (char -i 0)
#             | fzf --read0 --layout=reverse --height=40% -q (commandline)
#             | decode utf-8
#             | str trim
#         )"
#       }
#     ]
#   }  
  {
    name: insert_file
    modifier: alt
    keycode: char_y
    mode: [emacs vi_normal vi_insert]
    event: {
      send: executehostcommand
      cmd: "commandline --insert (fzf --tiebreak=chunk --layout=reverse  --multi --preview='echo {..}' --preview-window='bottom:3:wrap' --height=70% | decode utf-8 | str trim)"
    }
  }
  {
    name: insert_sudo
    modifier: alt
    keycode: char_s
    mode: [emacs, vi_insert, vi_normal]
    event: [
      { edit: MoveToStart }
      { send: ExecuteHostCommand,
        cmd: 'if (commandline | split row -r '\s+' | first) != `sudo` { commandline --insert `sudo `;commandline --cursor-end; }'
      }
    ]
  }
  ]
  # endregion
}
# endregion
# Custom Commands
source defs.nu

# region: prompts
# prompt
use "modules/prompt/oh-my.nu" git_prompt
$env.PROMPT_COMMAND = {|| (git_prompt).left_prompt }
$env.PROMPT_COMMAND_RIGHT = {|| (git_prompt).right_prompt }
$env.PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi green_bold)\n❯ (ansi reset)"} else {$"(ansi red_bold)\n❯ (ansi reset)"}}
$env.TRANSIENT_PROMPT_COMMAND = { "" }
$env.TRANSIENT_PROMPT_COMMAND_RIGHT = { || (git_prompt).right_prompt }
$env.TRANSIENT_PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi light_yellow_bold)❯ (ansi reset)"} else {$"(ansi light_red_bold)❯ (ansi reset)"}}

# use "modules/prompt/full-line.nu"
# $env.PROMPT_COMMAND = {|| full-line }
# $env.PROMPT_COMMAND_RIGHT = ""
# $env.PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi green_bold)\n❯ (ansi reset)"} else {$"(ansi red_bold)\n❯ (ansi reset)"}}

# $env.STARSHIP_SHELL = "nu"
# $env.STARSHIP_SESSION_KEY = (random chars -l 16)

# $env.PROMPT_INDICATOR = ""
# $env.PROMPT_MULTILINE_INDICATOR = (^starship prompt --continuation)

# $env.PROMPT_COMMAND = { ||
#     let width = (term size | get columns)
#     ^starship prompt $"--cmd-duration=($env.CMD_DURATION_MS)" $"--status=($env.LAST_EXIT_CODE)" $"--terminal-width=($width)"
# }

# $env.PROMPT_COMMAND_RIGHT = { ||
#     let width = (term size | get columns)
#     ^starship prompt --right $"--cmd-duration=($env.CMD_DURATION_MS)" $"--status=($env.LAST_EXIT_CODE)" $"--terminal-width=($width)"
# }

# source "prompt/oh-my-v2.nu"
# $env.PROMPT_COMMAND = { (get_prompt 24bit).left_prompt }
# $env.PROMPT_COMMAND_RIGHT = { (get_prompt 24bit).right_prompt }

# experiment with panache-git
# source "prompt/panache-git.nu"
# $env.PROMPT_COMMAND = { panache-git }
# endregions

# external completions
use "custom-completions/cargo/cargo-completions.nu" *
# use "custom-completions/git/git-completions.nu" *
# use "custom-completions/winget/winget-completions.nu" *

# git completions and aliases
use custom-completions/git/git-completions.nu *
use aliases/git/git-aliases.nu *

# temperature converter
use "sourced/temp.nu" *

# get_weather
use "modules/weather/get-weather.nu" get_weather

# zoxide
source ~/.zoxide.nu

# if LS_COLORS exists, hide it
if 'LS_COLORS' in ($env | columns) {
  hide-env LS_COLORS
}

# make sure my PATH is unique
$env.PATH = ($env.PATH | each {|r| $r | split row (char esep)} | flatten | uniq | str join (char esep))
# source /Users/fdncred/.config/broot/launcher/nushell/br

# alias ls = eza -lhA --color=always --group-directories-first --color-scale-mode=gradient --grid --icons --git --git-repos --color-scale all --no-user --no-permissions
```

ysthakur commented:
@fdncred Thanks! Tried your config and I still get that issue (Nushell 0.88), gonna give up on figuring that out :(

![image](https://github.com/nushell/reedline/assets/45539777/9b10a6bb-210c-4e71-9bcc-d455be2150ff)

ysthakur commented on src/menu/ide_menu.rs:
```diff
@@ -0,0 +1,1459 @@
+                cursor_pos as i16 + cursor_offset
+            } else {
+                (cursor_pos as i16 + cursor_offset)
+                    .saturating_sub(description_width as i16 + description_offset as i16)
```
There needs to be a check here to see if there's any description at all to avoid adding `description_offset` in that case. Should fix the jumping issue (fingers crossed)

ysthakur commented:
@fdncred In my config, it turns out I was sourcing Carapace, which spewed errors every time completions were triggered. When I stopped sourcing that, the jumping issue was fixed for me. Maybe you're also sourcing something like that?

maxomatic458 commented:
the problem is that whenever you trigger a completion with this prompt config, we send it from the second line (so thats where i get the cursor pos). But when the cursor teleports to the line above, no menu-event is being sent, so if you dont press anything the menu will think the cursor is where the actual completion was triggered from (line 2). Thats whats causing the jumping

maxomatic458 commented:
So id say the underlying problem is that this cursor teleportation is happening (without updating the menu)

ysthakur commented:
I don't know if sending another event would be possible but moving the update_working_details call to right after the prompt is drawn might help get the right cursor position?

maxomatic458 commented:
@ysthakur so i tried that but the problem is to refresh the cursor pos prior to rendering the full menu, the stdout needs to be flushed which leads to flickering in the terminal.

i think it would be better to update the cursor pos externally; so before getting the menu string; we could call something like menu.set_cursor_pos(...); 
and this could be done in [here](https://github.com/nushell/reedline/blob/ef7b96c157f644f97907858ec6aea6d377e6639b/src/painting/painter.rs#L230C5-L230C18)

stormasm commented:
https://github.com/nushell/reedline/issues/706

I reviewed the code and my only issue is the one I posted above...
Which is not really related to this PR per se --- but something to keep in mind
In the future prior to adding more menu types we should think about how to solve the above issue first 😄 

The nice thing about this PR is that all of new logic is located inside the *ide_menu.rs* so there is basically no risk of messing with any of the underlying reedline code...

I would say go ahead and land this PR once the other issues mentioned in the PR have been ironed out...

maxomatic458 commented:
@fdncred i think i have fixed it

it works for me now using your config
```
$env.PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi green_bold)\n❯ (ansi reset)"} else {$"(ansi red_bold)\n❯ (ansi reset)"}}
$env.TRANSIENT_PROMPT_COMMAND = ""
$env.TRANSIENT_PROMPT_INDICATOR = {|| if ($env.LAST_EXIT_CODE == 0) {$"(ansi light_yellow_bold)❯ (ansi reset)"} else {$"(ansi light_red_bold)❯ (ansi reset)"}}
```
could you give it a try now?

fdncred commented:
@maxomatic458 I'm trying to build your nushell main and am getting this. Is there a different way I should try this?
![image](https://github.com/nushell/reedline/assets/343840/6961277d-2652-40b2-a5f5-8406b900c4fd)

maxomatic458 commented:
@fdncred did you run ``cargo update``?

fdncred commented:
@maxomatic458 ok i did `cargo update -p reedline` and it compiles now but doesn't run. I guess it's because the my gstat plugin maybe?
```
Please join our Discord community at https://discord.gg/NtAbbGn
Our GitHub repository is at https://github.com/nushell/nushell
Our Documentation is located at https://nushell.sh
Tweet us at @nu_shell
Learn how to remove this at: https://nushell.sh/book/configuration.html#remove-welcome-message

It's been this long since Nushell's first commit:
4yrs 8months 7days 22hrs 56mins 52secs 295ms 948µs 900ns

Startup Time: 1sec 251ms 217µs

 .cargo    │  assets  │  devdocs │  src   │  .gitattributes │  Cargo.toml         │  Cross.toml │  rust-toolchain.toml
 .githooks │  benches │  docker  │  tests │  .gitignore     │  CODE_OF_CONDUCT.md │  LICENSE    │  toolkit.nu
 .github   │  crates  │  scripts │  wix   │  Cargo.lock     │  CONTRIBUTING.md    │  README.md  │
toolkit module detected...
activating toolkit module with `use toolkit.nu`
Error:   × Plugin failed to decode
  help: invalid length 3, expected struct CallInfo with 4 elements


Error:   × Plugin failed to decode
  help: invalid length 3, expected struct CallInfo with 4 elements


Error:   × Plugin failed to decode
  help: invalid length 3, expected struct CallInfo with 4 elements

```

I'm going to try cargo clean/rebuild.

fdncred commented:
oh, this is because your main may be out of date. we had a plugin change in the last week or so.

maxomatic458 commented:
oh ok, ill sync my fork

fdncred commented:
I think it's probably good enough to land but it's still weird that it doesn't start where my cursor is on the second line of my prompt. I just don't think that's something you've introduced, but has been there all along. So, I wouldn't expect you to fix that part in this PR.
![ide](https://github.com/nushell/reedline/assets/343840/7be8ba19-b4de-4394-acf3-75a85f3e778a)

maxomatic458 commented:
@fdncred 
yes, same thing happens using any menu (with your config)
i think that happens because your base indicator contains a ``\n``,
but your completion indicator doesnt

i think you could fix that by just adding a ``\n`` to your completion indicator (because the default indicator just gets replaced with the menu indicator) 
so i think its not a bug, but just a weird config

fdncred commented:
That's mostly it but something else is going on too but that's for another PR.
![image](https://github.com/nushell/reedline/assets/343840/4368c47d-249d-46ba-b991-72951bab0ae6)

fdncred commented:
Thanks for all your hard work on this PR!

fdncred commented:
Is there a PR in nushell to use `set_cursor_pos` in nushell so we can build with this PR? If not can you create one? Apparently, your PR is another PR https://github.com/nushell/nushell/pull/11535#issuecomment-1902077334 because it isn't implemented.

maxomatic458 commented:
ill make one right now

fdncred commented:
awe, thank you so much! we really appreciate it.

maxomatic458 commented:
@fdncred just made one https://github.com/nushell/nushell/pull/11589

fdncred commented:
@maxomatic458 one last thing. I was trying to use this ide style menu and i'm getting errors in nushell. What am I getting wrong?

I added this to my menus in config.nu
```
  {
    name: completion_menu
    only_buffer_difference: false
    marker: $" \n❯ (char -u '1f4ce') "
    type: {
      layout: ide # the new ide style menu
      border: true
    }
    style: {
      text: { fg: "#33ff00" }
      selected_text: { fg: "#0c0c0c" bg: "#33ff00" attr: b}
      description_text: yellow
    }
  }
```
![image](https://github.com/nushell/reedline/assets/343840/4078ac36-417b-4f80-848d-cad3c8198244)
Maybe there are some other changed needed in nushell that I'm not seeing because I'm half asleep still. 😆

maxomatic458 commented:
@fdncred yea there is no support currently for loading it from config.nu, its pretty easy to add (ive already done it in my fork, which i have deleted by accident)
i can make a PR for this later. But i think ill do this after https://github.com/nushell/reedline/pull/708 is merged (so multiline edits properly work)

fdncred commented:
ya, i was looking for your fork. LOL. i'm fine with waiting for 708. thanks!

fdncred commented:
Have you run into this? I can't seem to select a word+space, like " word" or "word " on my mac. I'm doing shift+ctrl+left and shift+ctrl+right in the gif. I could've swore i was able to do it on my windows machine.
![select](https://github.com/nushell/reedline/assets/343840/b0471354-11a6-470a-84e6-eadf31394be8)

maxomatic458 commented:
on windows im able to do this here:
" word"
cursor at the start, then ctrl+shift+right
will select the whole string (including the space at the start)


same works with  "word ", when the cursor is behind the the last space
do you think this might be related to this PR?

but i dont really see whats wrong in your clip,
do you mean that you would require an extra press of ctrl+shift+right to select the space as well?

fdncred commented:
on mac, i was able to figure out that shift+left was doing the selection character by character. doing ctrl+shift+left is doing the selection word by word.

what i was expecting was that ctrl+shift+left would select character by character vs word by word. maybe that's my mistake.

I'm also wondering if terminal bindings come into play here with selection, like we saw with select_all previously with WT.
